File: SymbolReplacer.java
class SymbolReplacer {
    protected String stringToReplace;
    protected List<String> alreadyReplaced = new ArrayList<String>();

    SymbolReplacer(String s) {
      this.stringToReplace = s;
    }

    String replace() {
      Pattern symbolPattern = Pattern.compile("\\$([a-zA-Z]\\w*)");
      Matcher symbolMatcher = symbolPattern.matcher(stringToReplace);
      while (symbolMatcher.find()) {
        String symbolName = symbolMatcher.group(1);
        if (getSymbol(symbolName) != null && !alreadyReplaced.contains(symbolName)) {
          alreadyReplaced.add(symbolName);
          stringToReplace = stringToReplace.replace("$" + symbolName, translate(symbolName));
        }
      }
      return stringToReplace;
    }

    protected String translate(String symbolName) {
      return getSymbol(symbolName);
    }
  }

File: lib/symbol_replacer.rb
class SymbolReplacer
  
    def initialize(s) 
      @string_to_replace = s
      @already_replaced = []
    end

    def replace
      @string_to_replace.scan(/\$([a-zA-Z]\w*)/).each do |match|
        global_name = match[0] 
        
        if !eval("$" + global_name).nil? && !@already_replaced.include?(global_name) then
          @already_replaced << global_name
          @string_to_replace = @string_to_replace.gsub("$" + global_name, translate(global_name));
        end
      end
      @string_to_replace
    end
    
    def translate(global_name)
      eval("$" + global_name)
    end
        
end
